File: soyeonLim/week02/BOJ_2470_두용액.py
import sys

n = int(sys.stdin.readline().strip())
l = list(map(int, sys.stdin.readline().strip().split(' ')))

l.sort() #-99, -2, -1, 4, 98

start = 0
end = n-1

answer = abs(l[start]+l[end])
final = [l[start],l[end]]

while start<end:
    left = l[start]
    right = l[end]

    sum = left+right

    if abs(sum)<answer:
        answer = abs(sum)
        final = [left,right]
        if answer ==0:
            break
    if sum<0:
        start+=1
    else:
        end-=1

print(final[0],final[1])

File: chansik/src/week02/BOJ_2470_두용액.java
package chansik.src.week02;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BOJ_2470_두용액 {
    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(bf.readLine());
        int n = Integer.parseInt(st.nextToken());
        int[] arr = new int[n];
        st = new StringTokenizer(bf.readLine());
        for(int i=0;i<n;i++) arr[i] = Integer.parseInt(st.nextToken());
        Arrays.sort(arr);

        int left = 0;
        int right = n-1;
        int pivot = Integer.MAX_VALUE;
        int min = Integer.MAX_VALUE;
        int max = Integer.MIN_VALUE;
        while(left < right) {
            int sum = Math.abs(arr[left] + arr[right]);

            if (pivot >= sum) {
                pivot = sum;
                min = arr[left];
                max = arr[right];
                if(pivot == 0) break;
            }

            if (arr[left] + arr[right] < 0) {
                left++;
            } else {
                right--;
            }
        }
        System.out.println(min + " " + max);
    }
}
